OpenRA.Mods.Common/Traits/Modifiers/FrozenUnderFog.cs

```diff
@@ -86,9 +98,14 @@ void ICreatesFrozenActors.OnVisibilityChanged(FrozenActor frozen)
 			if (!created)
 				return;
 
-			// Update state visibility to match the frozen actor to ensure consistency within the tick
-			// The rest of the state will be updated by ITick.Tick below
-			frozenStates[frozen.Viewer].IsVisible = !frozen.Visible;
+			// Update state visibility to match the frozen actor to ensure consistency
+			var playerIndex = frozen.Viewer.World.Players.IndexOf(frozen.Viewer);
+			var state = frozenStates[playerIndex];
```

Comment:
Fwiw, `frozenStates` can be indexed directly using `frozen.Viewer` (see the deleted code), so we could defer the `playerIndex` IndexOf-lookup to when it is actually needed (i.e., `isVisible` being true).

Reply:
Updated